### ISSUE TITLE
profile_mass guard #8 refuses all three line cases together; measured line fluxes are as linear in the mass amplitude as photometry

### ISSUE THREAD

**Issue by suchethac** [suchethac/tengri#2360]
Guard #8 refuses `profile_mass` whenever **any** emission-line channel is configured. The stated concern — that line fluxes may not be linear in the mass amplitude — is measurably false for the measured-line-flux case, and the prediction machinery needed to support it already exists.

## What the guard does today

`src/tengri/inference/mass_profile.py`:

```python
if fitter._fits_lines(fitter.model):
    return "an emission-line channel is configured (marginalized, fitted, or measured)", {}
```

`Fitter._fits_lines` is the OR of three unrelated situations:

| case | what it means | should it be refused? |
|---|---|---|
| `_eline_marginalize` | line amplitudes already analytically marginalized by the likelihood | **yes, for now** — two marginals interacting, needs its own analysis (see #2354) |
| `_eline_fitted` | line amplitudes are free parameters | **yes, for now** — the free amplitudes absorb the mass scale in the line channel |
| `_fits_line_fluxes` | **measured** line fluxes as a data channel | **no** — see below |

Only the third is a plain data channel with a mass-proportional prediction, and it is the one real user configuration: `notebooks/stochastic_sfh_recovery.py` case B, photometry plus 8 measured line fluxes.

## Emission lines are as linear in the mass amplitude as photometry is

Measured by the guard's own statistic, `max|prediction(mid+1 dex) / prediction(mid) - 10|`, on 7 bands (GALEX + SDSS), `neb=builders.neb.ssp()`, two-component Calzetti, `logzsol` and redshift fixed:

| SSP | SFH | guard's theta | worst of 12 prior draws |
|---|---|---|---|
| plain `fsps_prsc_miles` | `dpl` | 1.208e-13 | 1.652e-13 |
| **wNE** `prsc_miles_wNE` | `dpl` | 1.670e-13 | 1.901e-13 |
| plain `fsps_prsc_miles` | `dpl`+`field` | 3.682e-09 | 8.497e-06 |
| **wNE** `prsc_miles_wNE` | `dpl`+`field` | 3.651e-09 | 8.342e-06 |

Swapping a plain library for a wNE library — which bakes nebular line emission into the templates — changes the departure from linearity by less than a factor of 1.4. Whatever drives the `dpl`+`field` rows, it is **not** the lines: it is the stochastic SFH, and it affects photometry identically. That is filed separately as #2359 and is a prerequisite for this work, not part of it.

Measured directly on the line fluxes themselves, same model, sweeping mass across the full 7.0–12.5 dex prior, `flux / 10**logM` should be constant:

- Hα, [OIII], [SII], [OII]: agree to 3.2e-11 on a well-conditioned draw.
- **Hβ looks catastrophic (1.25e-05) and is an artifact.** Its normalized value is **negative** (−3.3e-27): Hβ measures near zero against stellar Balmer absorption, so the relative deviation is a small-denominator effect, not a nonlinearity. `line_measurement.py` already warns about exactly this in its "Continuum caveat (Balmer)" note. Any future check here must not compute a relative deviation against a line that measures through an absorption trough.

The structural reason lines are linear is visible in the code: `_line_flux_from_means` is `(feature mean − continuum mean) × c/λ² × Δλ / 4πd_L²`, with the continuum a linear interpolation between two side-band means. Every term scales with the mass amplitude, and the subtraction of two linear quantities is linear.

## The machinery already exists

`mass_profile._predict_full_vector` delegates to `loss_functions._build_prediction`, which **already accepts** `has_line_fluxes` and `measured_line_defs`. Profiling passes them as `False`/`None` and its docstring says why:

> Every emission-line / line-ratio / spectral-index channel is excluded by construction: this module's own guards refuse any fit carrying one, so `_build_prediction` is always called with every feature-channel flag off.

So the exclusion is a consequence of the guard, not a limitation of the assembly. The shared prediction path can already build a photometry-plus-lines vector — reusing it is what keeps the profiled statistics scoring the same vector the unprofiled likelihood does.

## What would need care

- **The data and noise vectors must match the prediction exactly.** `_profile_stats` forms `Σ f²/σ²` and `Σ d f/σ²` over the full data vector; if the line block's order or noise differs from what `_build_prediction` emits, the marginal is silently wrong. The joint photometry-then-spectrum order is pinned by `tests/regression/bug/test_bug_1366_joint_data_record.py`; lines need the same treatment.
- **Keep `marginalized` and `fitted` refused,** with a reason string that names which case it is refusing. The present string lists all three, so a user cannot tell why.
- **Acceptance must be the brute-force harness**, `_assert_profiled_matches_brute_force` extended to the line channel, not a new bespoke check.
- **#2359 first.** Line fits tend to use stochastic SFHs, which is the model class where single-theta linearity certification is least trustworthy. Admitting them through guard #8 while guard #14 still certifies from one theta widens exactly the wrong door.

## Not urgent

No current user is blocked: the one line-fitting notebook runs fine with profiling off, and the paper work that uses this module is photometry-only and has confirmed no dependency. This is a capability improvement, and the honest ordering is #2359 before this.

**Comment by suchethac** [comment]
**Correcting how I described one comparison, and adding a test-bed warning for whoever implements this.**

The table above contrasts a "plain" library against a wNE one, both built with `neb=builders.neb.ssp()`, and I described the difference as swapping the nebular library. That is not what the comparison was. Per #2362, `BakedInBackend.predict_nebular_sed` returns zeros unconditionally and never reads `ssp_data.nebular`, so `neb={'type': 'ssp'}` on a library with nothing baked in yields **no nebular emission at all**, silently.

I checked which of my grids are bare, using the Hα test from #2362 — peak flux within ±8 Å of 6562.8 over neighbouring continuum (6400–6500 and 6650–6750), youngest age bin, metallicity nearest solar:

| grid | Hα / continuum | |
|---|---|---|
| `ssp_prsc_miles_chabrier_wNE_logGasU-3.0_logGasZ0.0` | **2.93** | nebular present |
| `fsps_prsc_miles_chabrier` | **0.95** | bare |
| `bc03_pdva_stelib_chabrier` | **0.95** | bare |

So the "plain" rows carried zero nebular emission.

**The conclusion is unaffected, and the comparison is actually stronger than I claimed.** It is not "one nebular library against another" — it is **no nebular emission at all against nebular baked in and contributing**, and mass-linearity was unchanged between them (1.208e-13 against 1.670e-13). That is a cleaner demonstration that nebular emission does not break linearity in the mass amplitude than the comparison I thought I was making.

The direct line-flux measurements are unaffected for a different reason: they were all run on the **wNE** grid, where the lines are real (Hα/continuum 2.93). Those are the measurements that matter for this issue — line fluxes scaling by exactly 10.0000000000001 for a one-dex mass step, and line-flux-per-unit-SFR invariant to 1e-14.

**Warning for the implementation.** Do not build the test bed for guard #8 on a bare grid with `neb={'type': 'ssp'}`. Every line flux will be zero, every linearity check will pass trivially, and the tests will look green while exercising nothing. Use a **wNE** library, and assert the Hα/continuum ratio as a fixture precondition so the bed cannot silently go bare — this is the `ratchet-tests-must-not-hardcode-the-fix` failure mode in a different costume: a test whose subject is absent passes for the wrong reason.

**Comment by suchethac** [comment]
Refining the feasibility claim above, in both directions. I wrote that "the exclusion is a consequence of the guard, not a limitation of the assembly". That is true of the *prediction* but understates what else has to happen, and I should say where the real work is.

## The prediction and the data vector are separate things

`_build_prediction` returns two objects. Line fluxes go into the `prediction` **dict** under `"line_fluxes"`; they are *not* in `predicted`, the concatenated data vector that `mass_profile._predict_full_vector` returns. So passing `has_line_fluxes=True` would make the model compute line fluxes and then not score them, because `_profile_stats` forms its sums over `predicted` alone.

The profiled statistics would need the concatenation — photometry (and/or spectrum) followed by the line block, with the line block's own `line_flux_obs` and `line_flux_err` in the matching positions of the data and noise vectors.

## The good news: the line term is the same algebraic form

```python
chi2_lines = jnp.sum(
    standardized_residual(
        data_args["line_flux_obs"], model_lf, data_args["line_flux_err"]
    ) ** 2
)
e_lh = e_lh + 0.5 * chi2_lines
```

A plain Gaussian chi-square over its own vector with its own errors, added to the photometric term. That is exactly what analytic amplitude marginalization needs: `A = sum f^2/sigma^2` and `B = sum d f/sigma^2` are additive across independent Gaussian blocks, so a photometry block and a line block simply concatenate. No new math.

## The caveat, which determines the real scope

That snippet is the **censored fall-through**, and its own comment says so: *"The user-likelihood path composes these via `CompositeLikelihood`; the censored path needs them inlined since it bypasses the cohort."*

So on the ordinary path the line channel is scored by a `Likelihood` adapter, not by that inline chi-square, and the adapter is not guaranteed to be a plain Gaussian. `CalibrationELineMarginalizedLikelihood` (#2354) is exactly the case where it is not — it marginalizes the line amplitudes with a plug-in `a_hat`, which is a different object that the mass marginal cannot simply concatenate with.

**So the implementation must inspect the adapter rather than assume it.** Concretely:

- admit the measured-line-flux case **only** when the composed likelihood for the line block is a plain Gaussian chi-square over `(line_flux_obs, line_flux_err)`;
- keep refusing when the adapter marginalizes or fits the line amplitudes, with a reason naming which of the three cases fired;
- and pin that with a test, because a silent widening here would marginalize the mass against a line term whose likelihood is not the quadratic the algebra assumes — which would be wrong without being loud.

This is more work than "flip the flag", but less than "invent the math". The math is additive and already right; what is missing is a concatenated data vector and a check on which likelihood is actually in play.

**Comment by suchethac** [comment]
The gating measurement for the guard change, now that the machinery exists (#2372) and the deviation can be measured on the **concatenated** vector rather than argued about.

This is `_linearity_max_deviation`'s own nine-theta loop — prior median at `i=0` plus eight prior draws, seed 2359 — run twice per theta, once on the photometry-only vector and once on photometry+lines, with the real `_predict_full_vector`. wNE grid, baked-in nebular backend, 8 bands, three measured lines (Halpha, Hbeta, [OIII]5007), mass probed 9.75 -> 10.75 dex (the prior's bounds midpoint, which is where the shipping code pins the placeholder).

| theta | photometry only | photometry + lines | worst entry of the joined vector |
|---|---|---|---|
| 0 (prior median) | 5.151e-14 | 1.670e-13 | `Hbeta`, pred +3.257e-17 |
| 1 | 4.796e-14 | 4.796e-14 | `des_r` |
| 2 | 5.151e-14 | 5.151e-14 | `sdss_u` |
| 3 | 1.705e-13 | **1.705e-13** | `des_r` |
| 4 | 4.796e-14 | 4.796e-14 | `sdss_z` |
| 5 | 2.487e-14 | 8.171e-14 | `Halpha`, pred +2.217e-14 |
| 6 | 5.151e-14 | 1.315e-13 | `Hbeta`, pred +8.493e-15 |
| 7 | 9.592e-14 | 9.592e-14 | `sdss_g` |
| 8 | 3.553e-14 | 8.171e-14 | `Halpha`, pred +2.551e-14 |
| **worst of nine** | **1.705e-13** | **1.705e-13** | a photometry band |

**Adding the line block does not move the worst-of-nine at all.** The maximum over the joined vector is set by a photometry band (`des_r` at theta 3), and the line entries — while they are the per-theta worst in four of the nine — sit at 8e-14 to 1.7e-13, the same roundoff scale as the bands beside them. Against `_LINEARITY_TOL = 1e-2` the guard **engages**, with about eleven orders of margin.

That is a considerably stronger statement than the single-theta numbers in the issue body: it is the worst of nine thetas, on the concatenated vector, through the shipping predictor, on the `measure_line_fluxes` branch (this configuration has no discrete line catalog, so it takes the continuum-subtracted path — the one with the cancellation risk, not the easy one).

## The magnitude floor is not needed, and that is now measured rather than assumed

I flagged earlier that a line measured through a stellar Balmer absorption trough could pollute a relative deviation, and that the guard might need a per-block magnitude floor. Two findings retire that:

- The originally measured Hbeta artifact had a **negative** normalized value (-3.3e-27). The probe's validity mask is `isfinite(pred) & (pred > 0.0)`, so it was already excluded. The measurement never demanded a floor; I had mis-remembered the mask as `!= 0`.
- On this configuration Hbeta is in **emission** (+3.257e-17 at the prior median, +8.493e-15 at theta 6), not absorption, and its deviation is roundoff-scale. Across all nine thetas no entry landed small-and-positive in a way that inflated the ratio.

So the guard change adds no floor. If a small-positive trough case ever does appear it will show up as a refusal with the measured deviation in the reason string — which #2370 made visible precisely so this kind of thing is diagnosable rather than mysterious — and the fix can then be justified by the case that produced it.

## What remains for the guard change

Unchanged from the scoping in #2372, and none of it is affected by the above:

- split guard #8 into three separately-named refusals (`_eline_marginalize` / `_eline_fitted` / a measured channel that is censored or owned by a user-supplied likelihood);
- check `LineFluxData.limit_mask`, which the existing censored guard cannot see because it reads `fitter.data_mask` (the photometry/spectroscopy vector only);
- pass the block at all six `_profile_stats` call sites, since sites 1085/1146/1194 define the posterior that is sampled and 1420 draws the reinserted mass from its conditional — a mismatch there is a reinserted mass from a different conditional, silent and with no shape error;
- exercise both line-flux predictors, since `_LineFluxBlock.waves` is only read on the `predict_line_fluxes` branch and nothing currently covers it.
